File: server/src/tests/repositories/inMemoryParticipantRepository.ts
import { Participant } from '@application/entities/participant';

import {
  IGetConversationsByUserIdResponse,
  IGetParticipantsByConversationIdResponse,
  ParticipantRepository,
} from '@application/repositories/participant-repository';
import { UserRepository } from '@application/repositories/user-repository';
import { GroupRepository } from '@application/repositories/group-repository';
import { ConversationRepository } from '@application/repositories/conversation-repository';

import { Either, left, right } from '@helpers/Either';

export class InMemoryParticipantRepository implements ParticipantRepository {
  public items: Participant[];

  public userRepository: UserRepository | null = null;
  public groupRepository: GroupRepository | null = null;
  public conversationRepository: ConversationRepository | null = null;

  constructor() {
    this.items = [];
  }

  setConversationRepository(conversationRepository: ConversationRepository) {
    this.conversationRepository = conversationRepository;
  }
  setUserRepository(userRepository: UserRepository) {
    this.userRepository = userRepository;
  }
  setGroupRepository(groupRepository: GroupRepository) {
    this.groupRepository = groupRepository;
  }

  async add(participant: Participant): Promise<Either<Error, void>> {
    this.items.push(participant);

    return right(void 0);
  }

  async remove(participant: Participant): Promise<Either<Error, void>> {
    this.items = this.items.filter(
      item => item.userId === participant.userId && item.conversationId === participant.conversationId,
    );

    return right(void 0);
  }

  async getByConversationId(conversationId: string): Promise<Either<Error, Participant[]>> {
    const participants = this.items.filter(item => item.conversationId === conversationId);

    return right(participants);
  }

  async getByUserId(userId: string): Promise<Either<Error, Participant[]>> {
    const userConversations = this.items.filter(item => item.userId === userId);

    return right(userConversations);
  }

  async getConversationsByUserId(userId: string): Promise<Either<Error, IGetConversationsByUserIdResponse[]>> {
    if (!this.conversationRepository) {
      return left(new Error('Conversation Repository must be passed to this test'));
    }
    if (!this.groupRepository) {
      return left(new Error('Group Repository must be passed to this test'));
    }

    const participationInConversations = this.items.filter(item => item.userId === userId);

    const conversationsWithParticipants = await Promise.all(
      participationInConversations.map(async participant => {
        const conversation = await this.conversationRepository!.getById(participant.conversationId);

        if (conversation.isLeft() || !conversation.value) {
          throw new Error('Invalid conversation query');
        }

        const group = await this.groupRepository!.findById(conversation.value.groupId || '');

        if (group.isLeft()) {
          throw new Error('Invalid group query');
        }

        return {
          id: conversation.value.id,
          createdAt: conversation.value.createdAt,
          disabledAt: conversation.value.disabledAt,
          isGroup: conversation.value.isGroup,
          groupDescription: group.value?.description || null,
          groupName: group.value?.name || null,
        };
      }),
    );

    return right(conversationsWithParticipants);
  }

  async getParticipantsByConversationId(
    conversationId: string,
  ): Promise<Either<Error, IGetParticipantsByConversationIdResponse[]>> {
    if (!this.userRepository) {
      return left(new Error('User Repository must be passed to this test'));
    }

    const participantsOfConversation = this.items.filter(item => item.conversationId === conversationId);

    const participants = await Promise.all(
      participantsOfConversation.map(async participant => {
        const user = await this.userRepository!.findById(participant.userId);

        if (user.isLeft() || !user.value) {
          throw new Error('Invalid user query');
        }

        const result: IGetParticipantsByConversationIdResponse = {
          id: user.value.id,
          name: user.value.name,
          email: user.value.email,
          lastOnline: user.value.lastOnline,
          enteredAt: participant.createdAt,
          removedAt: participant.removedAt,
        };

        return result;
      }),
    );

    return right(participants);
  }
}

File: Proxymity-Chat-Service/tests/ProxyMity.Tests/InMemoryRepositories/InMemoryParticipantRepository.cs
﻿namespace ProxyMity.Unit.InMemoryRepositories;

internal class InMemoryParticipantRepository(
    IGroupRepository groupRepository,
    IUserRepository userRepository,
    IConversationRepository conversationRepository
) : InMemoryRepository<Participant>, IParticipantRepository {

    public Task AddAsync(Participant participant, CancellationToken cancellationToken) {
        Items.Add(participant);
        return Task.CompletedTask;
    }

    public Task<List<Participant>> GetByConversationIdAsync(Ulid conversationId, CancellationToken cancellationToken) {
        var participants = Items.Where(p => p.ConversationId == conversationId).ToList();
        return Task.FromResult(participants);
    }

    public async Task<Participant?> GetByIdAsync(Ulid userId, Ulid conversationId, CancellationToken cancellationToken) {
        var participant = Items.SingleOrDefault(p => p.UserId == userId && p.ConversationId == conversationId);
        await Task.Run(() => { });

        return participant;
    }

    public Task<List<Participant>> GetByUserIdAsync(Ulid userId, CancellationToken cancellationToken) {
        var participants = Items.Where(p => p.UserId == userId).ToList();

        return Task.FromResult(participants);
    }

    public async Task<List<GetConversationsByUserIdQuery>> GetConversationsByUserIdAsync(Ulid userId, CancellationToken cancellationToken) {
        var participationInConversations = Items.FindAll(p => p.UserId == userId);
        var conversationsWithParticipants = new List<GetConversationsByUserIdQuery>();

        foreach (var item in participationInConversations) {
            Conversation conversation = await conversationRepository.GetByIdAsync(item.ConversationId, cancellationToken)
                ?? throw new ConversationNotFoundException(item.ConversationId);

            Group? group = await groupRepository.FindByIdAsync(conversation?.GroupId ?? Ulid.NewUlid(), cancellationToken);

            conversationsWithParticipants.Add(new GetConversationsByUserIdQuery(
                Id: conversation!.Id,
                CreatedAt: conversation.CreatedAt,
                GroupName: group?.Name,
                GroupDescription: group?.Description,
                GroupId: conversation?.GroupId,
                ConversationPinnedAt: item.ConversationPinnedAt
            ));
        }

        return conversationsWithParticipants;
    }

    public async Task<List<GetParticipantsByConversationIdQuery>> GetParticipantsByConversationIdAsync(Ulid conversationId, CancellationToken cancellationToken) {
        var participantsOfConversation = Items.FindAll(x => x.ConversationId == conversationId);

        var participants = new List<GetParticipantsByConversationIdQuery>();

        foreach (var item in participantsOfConversation) {
            var user = await userRepository.FindByIdAsync(item.UserId, cancellationToken);

            if (user is null) {
                continue;
            }

            var result = new GetParticipantsByConversationIdQuery(
                user.Id,
                user.FirstName,
                user.LastName,
                user.Email,
                user.PhotoUrl,
                user.LastOnline,
                item.CreatedAt,
                item.RemovedAt
            );

            participants.Add(result);
        }

        return participants;
    }

    public Task PinConversation(Ulid conversationId, Ulid userId, CancellationToken cancellationToken) {
        foreach (var item in Items) {
            if (item.UserId == userId && item.ConversationId == conversationId) {
                item.ConversationPinnedAt = DateTime.UtcNow;
            }
        }

        return Task.CompletedTask;
    }

    public Task Remove(Participant participant, CancellationToken cancellationToken) {
        Items.Remove(participant);

        return Task.CompletedTask;
    }

    public Task UnpinConversation(Ulid conversationId, Ulid userId, CancellationToken cancellationToken) {
        foreach (var item in Items) {
            if (item.UserId == userId && item.ConversationId == conversationId) {
                item.ConversationPinnedAt = null;
            }
        }

        return Task.CompletedTask;
    }
}
